### PR TITLE
Add JWT credentials

### PR DESCRIPTION
**Pull request by jonparrott**
~~(Has #8 as a diffbase)~~

### CODE REVIEW

File: google/auth/jwt.py [hunk 1]
```diff
@@ -234,3 +239,239 @@ def decode(token, certs=None, verify=True, audience=None):
...
     return payload
+
+
+class Credentials(credentials.Signing,
+                  credentials.Credentials):
+    """Credentials that use a JWT as the bearer token.
+
+    The constructor arguments determine the claims for the JWT that is
+    sent with requests. Usually, you'll construct these credentials with
+    one of the helper constructors.
```
Comment (dhermes) [inline google/auth/jwt.py]: Name the helper constructors?

File: google/auth/jwt.py [hunk 2]
```diff
@@ -234,3 +239,239 @@ def decode(token, certs=None, verify=True, audience=None):
...
+    The claims are considered immutable. If you want to modify the claims,
+    you can easily create another instance using :meth:`with_claims`::
+
+        new_credentials = credentials.with_claims(
+            audience='https://vision.googleapis.com')
+
+    Note that JWT credentials will also set the audience claim on demand. If no
+    audience is specified when creating the credentials, then whenever a
+    request is made the credentials will automatically generate a one-time
+    JWT with the request URI as the audience.
```
Comment (dhermes) [inline google/auth/jwt.py]: Shouldn't this be the first thing you say? And shouldn't there be an example?

File: google/auth/jwt.py [hunk 3]
```diff
@@ -234,3 +239,239 @@ def decode(token, certs=None, verify=True, audience=None):
...
+            key_id = info['private_key_id']
+            private_key = info['private_key']
+        except KeyError:
+            raise ValueError(
+                'Service account info was not in the expected format.')
+
+        signer = crypt.Signer.from_string(private_key, key_id)
+
+        kwargs.setdefault('subject', email)
+        return cls(signer, issuer=email, **kwargs)
```
Comment (dhermes) [inline google/auth/jwt.py]: I feel like #25 and this PR share a lot of code. Not worth trying to consolidate?

File: google/auth/jwt.py [hunk 4]
```diff
@@ -234,3 +239,239 @@ def decode(token, certs=None, verify=True, audience=None):
...
+
+        Returns:
+            google.auth.jwt.Credentials: A new credentials instance.
+        """
+        return Credentials(
+            self._signer,
+            issuer=issuer if issuer is not None else self._issuer,
+            subject=subject if subject is not None else self._subject,
+            audience=audience if audience is not None else self._audience,
+            additional_claims=dict(self._additional_claims).update(
```
Comment (dhermes) [inline google/auth/jwt.py]: `dict()` in lieu of `.copy()`?

File: google/auth/jwt.py [hunk 5]
```diff
@@ -234,3 +239,239 @@ def decode(token, certs=None, verify=True, audience=None):
...
+        now = _helpers.utcnow()
+        lifetime = datetime.timedelta(seconds=self._token_lifetime)
+        expiry = now + lifetime
+
+        payload = {
+            'iss': self._issuer,
+            'sub': self._subject or self._issuer,
+            'iat': _helpers.datetime_to_secs(now),
+            'exp': _helpers.datetime_to_secs(expiry),
+            'aud': audience or self._audience
```
Comment (dhermes) [inline google/auth/jwt.py]: Trailing comma

File: google/auth/jwt.py [hunk 6]
```diff
@@ -234,3 +239,239 @@ def decode(token, certs=None, verify=True, audience=None):
...
+
+    def refresh(self, request):
+        """Refreshes the access token.
+
+        Args:
+            request (Any): Unused.
+        """
+        # pylint: disable=unused-argument
+        # (pylint doens't correctly recognize overriden methods.)
+
```
Comment (dhermes) [inline google/auth/jwt.py]: Empty line just for fun? No corresponding `pylint: enable`? Shouldn't the unused argument be before `refresh()` is defined?

File: google/auth/jwt.py [hunk 7]
```diff
@@ -234,3 +239,239 @@ def decode(token, certs=None, verify=True, audience=None):
...
+        token for the request URI. In either case, it will set the
+        authorization header in headers to the token.
+
+        Args:
+            request (Any): Unused.
+            method (str): The request's HTTP method.
+            url (str): The request's URI.
+            headers (Mapping): The request's headers.
+        """
+        # pylint: disable=unused-argument
```
Comment (dhermes) [inline google/auth/jwt.py]: No corresponding `pylint: enable`. Also s/doens't/doesn't

File: tests/data/service_account.json [hunk 8]
```diff
@@ -0,0 +1,10 @@
+{
```
Comment (dhermes) [inline tests/data/service_account.json]: Is this file also added in another PR?

File: tests/test_jwt.py [hunk 9]
```diff
@@ -187,3 +194,143 @@ def test_roundtrip_explicit_key_id(token_factory):
     certs = {'2': OTHER_CERT_BYTES, '3': PUBLIC_CERT_BYTES}
     payload = jwt.decode(token, certs)
     assert payload['user'] == 'billy bob'
+
+
+class TestCredentials:
+    service_account_email = 'service-account@example.com'
```
Comment (dhermes) [inline tests/test_jwt.py]: UPPERCASE please

File: tests/test_jwt.py [hunk 10]
```diff
@@ -187,3 +194,143 @@ def test_roundtrip_explicit_key_id(token_factory):
...
+
+    @pytest.fixture(autouse=True)
+    def credentials(self, signer):
+        self.credentials = jwt.Credentials(
```
Comment (dhermes) [inline tests/test_jwt.py]: Please rename `def credentials`

File: tests/test_jwt.py [hunk 11]
```diff
@@ -187,3 +194,143 @@ def test_roundtrip_explicit_key_id(token_factory):
...
+            info = json.load(fh)
+
+        credentials = jwt.Credentials.from_service_account_info(info)
+
+        assert credentials._signer.key_id == info['private_key_id']
+        assert credentials._issuer == info['client_email']
+        assert credentials._subject == info['client_email']
+
+    def test_from_service_account_info_args(self):
+        info = dict(SERVICE_ACCOUNT_INFO)
```
Comment (dhermes) [inline tests/test_jwt.py]: `dict()` in lieu of `.copy()`?

File: tests/test_jwt.py [hunk 12]
```diff
@@ -187,3 +194,143 @@ def test_roundtrip_explicit_key_id(token_factory):
...
+
+        credentials = jwt.Credentials.from_service_account_info(
+            info, subject='subject', audience='audience',
+            additional_claims={'meta': 'data'})
+
+        assert credentials._signer.key_id == info['private_key_id']
+        assert credentials._issuer == info['client_email']
+        assert credentials._subject == 'subject'
+        assert credentials._audience == 'audience'
+        assert credentials._additional_claims['meta'] == 'data'
```
Comment (dhermes) [inline tests/test_jwt.py]: Have `subject`, `audience` etc. hardcoded in multiple places. DRY?

File: tests/test_jwt.py [hunk 13]
```diff
@@ -187,3 +194,143 @@ def test_roundtrip_explicit_key_id(token_factory):
...
+
+    def test_from_service_account_bad_key(self):
+        info = dict(SERVICE_ACCOUNT_INFO)
```
Comment (dhermes) [inline tests/test_jwt.py]: `.copy()`?

File: tests/test_jwt.py [hunk 14]
```diff
@@ -187,3 +194,143 @@ def test_roundtrip_explicit_key_id(token_factory):
...
+            info, subject='subject', audience='audience',
+            additional_claims={'meta': 'data'})
+
+        assert credentials._signer.key_id == info['private_key_id']
+        assert credentials._issuer == info['client_email']
+        assert credentials._subject == 'subject'
+        assert credentials._audience == 'audience'
+        assert credentials._additional_claims['meta'] == 'data'
+
+    def test_from_service_account_bad_key(self):
```
Comment (dhermes) [inline tests/test_jwt.py]: When I saw `bad_key` I originally assumed you meant bad dictionary key. How about `bad_private_key`?

File: tests/test_jwt.py [hunk 15]
```diff
@@ -187,3 +194,143 @@ def test_roundtrip_explicit_key_id(token_factory):
...
+            jwt.Credentials.from_service_account_info(info)
+
+        assert excinfo.match(r'No key could be detected')
+
+    def test_from_service_account_bad_format(self):
+        with pytest.raises(ValueError):
+            jwt.Credentials.from_service_account_info({})
+
+    def test_from_service_account_file(self):
+        info = dict(SERVICE_ACCOUNT_INFO)
```
Comment (dhermes) [inline tests/test_jwt.py]: `.copy()`?

File: tests/test_jwt.py [hunk 16]
```diff
@@ -187,3 +194,143 @@ def test_roundtrip_explicit_key_id(token_factory):
...
+
+        credentials = jwt.Credentials.from_service_account_file(
+            SERVICE_ACCOUNT_JSON_FILE)
+
+        assert credentials._signer.key_id == info['private_key_id']
+        assert credentials._issuer == info['client_email']
+        assert credentials._subject == info['client_email']
+
+    def test_from_service_account_file_args(self):
+        info = dict(SERVICE_ACCOUNT_INFO)
```
Comment (dhermes) [inline tests/test_jwt.py]: `.copy()`?

File: tests/test_jwt.py [hunk 17]
```diff
@@ -187,3 +194,143 @@ def test_roundtrip_explicit_key_id(token_factory):
...
+
+        credentials = jwt.Credentials.from_service_account_file(
+            SERVICE_ACCOUNT_JSON_FILE, subject='subject', audience='audience',
+            additional_claims={'meta': 'data'})
+
+        assert credentials._signer.key_id == info['private_key_id']
+        assert credentials._issuer == info['client_email']
+        assert credentials._subject == 'subject'
+        assert credentials._audience == 'audience'
+        assert credentials._additional_claims['meta'] == 'data'
```
Comment (dhermes) [inline tests/test_jwt.py]: Hardcode?

File: tests/test_jwt.py [hunk 18]
```diff
@@ -187,3 +194,143 @@ def test_roundtrip_explicit_key_id(token_factory):
...
+
+    def test_default_state(self):
+        assert not self.credentials.valid
+        # Expiration hasn't been set yet
+        assert not self.credentials.expired
+
+    def test_sign_bytes(self):
+        to_sign = b'123'
+        signature = self.credentials.sign_bytes(to_sign)
+        crypt.verify_signature(to_sign, signature, PUBLIC_CERT_BYTES)
```
Comment (dhermes) [inline tests/test_jwt.py]: No actual asserts here. How could you test something, other than a successful exit?

File: tests/test_jwt.py [hunk 19]
```diff
@@ -187,3 +194,143 @@ def test_roundtrip_explicit_key_id(token_factory):
...
+        assert not self.credentials.expired
+
+    def test_expired(self):
+        assert not self.credentials.expired
+
+        self.credentials.refresh(None)
+        assert not self.credentials.expired
+
+        with mock.patch('google.auth._helpers.utcnow') as now:
+            one_day_from_now = datetime.timedelta(days=1)
```
Comment (dhermes) [inline tests/test_jwt.py]: To be pedantic, a delta can't be one day from now, it's just the from now part.

File: tests/test_jwt.py [hunk 20]
```diff
@@ -187,3 +194,143 @@ def test_roundtrip_explicit_key_id(token_factory):
...
+
+    def test_before_request_one_time_token(self):
+        headers = {}
+
+        self.credentials.refresh(None)
+        self.credentials.before_request(
+            mock.Mock(), 'GET', 'http://example.com?a=1#3', headers)
+
+        header_value = headers['authorization']
+        token = header_value.split().pop()
```
Comment (dhermes) [inline tests/test_jwt.py]: The `pop()` is uncomfortable to me. Can you very what's left after the `pop()`?

File: tests/test_jwt.py [hunk 21]
```diff
@@ -187,3 +194,143 @@ def test_roundtrip_explicit_key_id(token_factory):
...
+    def test_before_request_set_audience(self):
+        headers = {}
+
+        credentials = self.credentials.with_claims(audience='test')
+        credentials.refresh(None)
+        credentials.before_request(
+            None, 'GET', 'http://example.com?a=1#3', headers)
+
+        header_value = headers['authorization']
+        token = header_value.split().pop()
```
Comment (dhermes) [inline tests/test_jwt.py]: Ditto on `pop()`

File: tests/test_jwt.py [hunk 22]
```diff
@@ -187,3 +194,143 @@ def test_roundtrip_explicit_key_id(token_factory):
...
+        payload = self._verify_token(token)
+        assert payload['aud'] == 'http://example.com'
+
+    def test_before_request_set_audience(self):
+        headers = {}
+
+        credentials = self.credentials.with_claims(audience='test')
+        credentials.refresh(None)
+        credentials.before_request(
+            None, 'GET', 'http://example.com?a=1#3', headers)
```
Comment (dhermes) [inline tests/test_jwt.py]: Is this `None` the same `None` in `refresh()`?

File: google/auth/jwt.py (hunk 1 above)
Comment (jonparrott) [inline google/auth/jwt.py]: (discussed offline, resolved to leave as-is)

File: google/auth/jwt.py (hunk 2 above)
Comment (jonparrott) [inline google/auth/jwt.py]: Done.

File: google/auth/jwt.py (hunk 3 above)
Comment (jonparrott) [inline google/auth/jwt.py]: Probably. Filed #27 to handle this after the fact.

File: google/auth/jwt.py (hunk 4 above)
Comment (jonparrott) [inline google/auth/jwt.py]: py2.6-isms dying hard. Fixed.

File: google/auth/jwt.py (hunk 5 above)
Comment (jonparrott) [inline google/auth/jwt.py]: Done.

File: google/auth/jwt.py (hunk 6 above)
Comment (jonparrott) [inline google/auth/jwt.py]: Remove empty new line.

> Shouldn't the unused argument be before refresh() is defined?

Apparently not.

File: google/auth/jwt.py (hunk 7 above)
Comment (jonparrott) [inline google/auth/jwt.py]: Fixed typo. unused-argument applies to the whole block, so there's no need to disable it (unless I defined an inner function).

File: tests/data/service_account.json (hunk 8 above)
Comment (jonparrott) [inline tests/data/service_account.json]: Yes.

File: tests/test_jwt.py (hunk 9 above)
Comment (jonparrott) [inline tests/test_jwt.py]: Done.

File: tests/test_jwt.py (hunk 10 above)
Comment (jonparrott) [inline tests/test_jwt.py]: Done.

File: tests/test_jwt.py (hunk 11 above)
Comment (jonparrott) [inline tests/test_jwt.py]: Done.

File: tests/test_jwt.py (hunk 12 above)
Comment (jonparrott) [inline tests/test_jwt.py]: Done.

File: tests/test_jwt.py (hunk 17 above)
Comment (jonparrott) [inline tests/test_jwt.py]: Done.

File: tests/test_jwt.py (hunk 14 above)
Comment (jonparrott) [inline tests/test_jwt.py]: Done.

File: tests/test_jwt.py (hunk 13 above)
Comment (jonparrott) [inline tests/test_jwt.py]: Fixed.

File: tests/test_jwt.py (hunk 15 above)
Comment (jonparrott) [inline tests/test_jwt.py]: Fixed.

File: tests/test_jwt.py (hunk 16 above)
Comment (jonparrott) [inline tests/test_jwt.py]: Fixed.

File: tests/test_jwt.py (hunk 18 above)
Comment (jonparrott) [inline tests/test_jwt.py]: Fixed.

File: tests/test_jwt.py (hunk 19 above)
Comment (jonparrott) [inline tests/test_jwt.py]: Lol Fixed.

File: tests/test_jwt.py (hunk 20 above)
Comment (jonparrott) [inline tests/test_jwt.py]: Fixed by specifying an argument to 'split'.

File: google/auth/jwt.py (hunk 6 above)
Comment (dhermes) [inline google/auth/jwt.py]: @jonparrott This is still pending

File: tests/test_jwt.py (hunk 22 above)
Comment (jonparrott) [inline tests/test_jwt.py]: Yes, unused `request`.

File: tests/test_jwt.py (hunk 21 above)
Comment (jonparrott) [inline tests/test_jwt.py]: Fixed likewise.

File: tests/test_jwt.py [hunk 23]
```diff
@@ -197,12 +197,16 @@ def test_roundtrip_explicit_key_id(token_factory):
 
 
 class TestCredentials:
-    service_account_email = 'service-account@example.com'
+    SERVICE_ACCOUNT_EMAIL = 'service-account@example.com'
+    SUBJECT = 'subject'
+    AUDIENCE = 'audience'
+    ADDITIONAL_CLAIMS = {'meta': 'data'}
+    credentials = None
```
Comment (dhermes) [inline tests/test_jwt.py]: Was putting `credentials` here just a reflex or was it required by Pylint?

File: tests/test_jwt.py [hunk 24]
```diff
@@ -302,7 +306,7 @@ def test_before_request_one_time_token(self):
             mock.Mock(), 'GET', 'http://example.com?a=1#3', headers)
 
         header_value = headers['authorization']
-        token = header_value.split().pop()
+        token = header_value.split('Bearer ').pop()
```
Comment (dhermes) [inline tests/test_jwt.py]: Still not happy here. Do the split on its own, then pop the token, then verify that `parts == []` after popping

File: tests/test_jwt.py (hunk 23 above)
Comment (jonparrott) [inline tests/test_jwt.py]: Pylint.

File: tests/test_jwt.py (hunk 24 above)
Comment (jonparrott) [inline tests/test_jwt.py]: Fixed.

**Comment by dhermes** [comment]
LGTM

**Pull request merged**